### PR TITLE
Improve index.d.ts for NodeNext

### LINKED ISSUE #3264: Incorrect exports for TypeScript in ESM mode
**Expected behaviour**
`import tracer from "dd-trace";` works in TypeScript ESM mode

**Actual behaviour**
Instead of tracer referring to the tracer, an extra `.default` access is required.

**Steps to reproduce**
See https://arethetypeswrong.github.io/?p=dd-trace%404.2.0 for a description of the issue and suggested fix. A TS project with moduleResolution set to "NodeNext" and package.json's "type" set to "module" reproduces it.

**Environment**

* **Operation system:** macOS 13.4
* **Node.js version:** node 18.x
* **Tracer version:** 4.2.0
* **Agent version:** –
* **Relevant library versions:** –

### PR DESCRIPTION
**Pull request by benasher44**
### What does this PR do?
Fixes #3264

I moves all types, except Tracer and Plugins, into the namespace and switches to `export =` to fix compatibility with nodenext

### Motivation
Fix support for NodeNext (newer TypeScript module resolution)

### Plugin Checklist

- [ ] Unit tests.
- [x] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Security 
Datadog employees:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

**Comment by benasher44** [comment]
@khanayan123 this should be semver-patch, but I don't have the ability to add labels.

Also this diff is large, but it's pretty easy to review if you hide whitespace changes:

image

**Comment by benasher44** [comment]
@khanayan123 maybe this could be in the v5 preview — people can test it out 😅 🙏

File: index.d.ts [hunk 1]
```diff
@@ -83,9 +84,9 @@ export declare interface Tracer extends opentracing.Tracer {
    * unless there is already an active span or `childOf` option. Note that this
    * option is deprecated and has been removed in version 4.0.
    */
-  trace<T> (name: string, fn: (span: Span) => T): T;
-  trace<T> (name: string, fn: (span: Span, done: (error?: Error) => void) => T): T;
-  trace<T> (name: string, options: TraceOptions & SpanOptions, fn: (span?: Span, done?: (error?: Error) => void) => T): T;
+  trace<T> (name: string, fn: (span?: tracer.Span, fn?: (error?: Error) => any) => T): T;
+  trace<T> (name: string, options: tracer.TraceOptions & tracer.SpanOptions, fn: (span?: tracer.Span, done?: (error?: Error) => string) => T): T;
```
Comment (Qard) [inline index.d.ts:88]: This change is incorrect. There is no done function which returns a string.

File: index.d.ts [hunk 2]
```diff
@@ -83,9 +84,9 @@ export declare interface Tracer extends opentracing.Tracer {
    * unless there is already an active span or `childOf` option. Note that this
    * option is deprecated and has been removed in version 4.0.
    */
-  trace<T> (name: string, fn: (span: Span) => T): T;
-  trace<T> (name: string, fn: (span: Span, done: (error?: Error) => void) => T): T;
-  trace<T> (name: string, options: TraceOptions & SpanOptions, fn: (span?: Span, done?: (error?: Error) => void) => T): T;
+  trace<T> (name: string, fn: (span?: tracer.Span, fn?: (error?: Error) => any) => T): T;
```
Comment (Qard) [inline index.d.ts:87]: This was split into two signatures both of which _require_ that span is handled.

**Review by Qard — CHANGES_REQUESTED** [review]
Looks like it's mostly just the change of `export default tracer` to `export = tracer` that matters here? What's the reason for all the other changes? Are those necessary? I worry a bit that so much structural changes might make this need to be a semver major change. 🤔 

Also, looks like a few things changed that should not.

**Comment by benasher44** [comment]
Ah whoops will fix those issues — thanks for flagging!

You're right that's the main focus, but in switching to `export =`, it cascades (the reason for the other changes). Once you have `export =`, that has the be the sole export (TS errors). To then support all of the other named exports, you move them into the namespace and re-export them there. In doing that, you then have to remove all of the `declare`, since declare is only valid/necessary at the top level.

File: index.d.ts [hunk 3]
```diff
@@ -117,1865 +118,1869 @@ export declare interface Tracer extends opentracing.Tracer {
...
-   * Returns the string representation of the internal trace ID.
-   */
-  toTraceId (): string;
-
-  /**
-   * Returns the string representation of the internal span ID.
-   */
-  toSpanId (): string;
+declare namespace tracer {
+  export type SpanOptions = opentracing.SpanOptions;
```
Comment (benasher44) [inline index.d.ts:131]: This is done this way to force TS to see this as a type import and not a value import (which isn't allowed in a namespace)

Comment (benasher44) [reply index.d.ts:131]: (as opposed to including it in the below export)

File: index.d.ts [hunk 4]
```diff
@@ -117,1865 +118,1871 @@ export declare interface Tracer extends opentracing.Tracer {
...
-}
+    /**
+     * Scopes or granted authorizations the user currently possesses.
+     * The value could come from the scope associated with an OAuth2
+     * Access Token or an attribute value in a SAML 2 Assertion.
+     */
+    scope?: string,
 
-/** @hidden */
-interface Plugins {
```
Comment (benasher44) [inline index.d.ts:824]: moved to top level to avoid exporting (see comment)

File: index.d.ts [hunk 5]
```diff
@@ -117,1865 +118,1871 @@ export declare interface Tracer extends opentracing.Tracer {
...
+    experimental?: {
+      b3?: boolean
+      traceparent?: boolean
+
       /**
-       * Whether to enable vulnerability deduplication
+       * Whether to add an auto-generated `runtime-id` tag to metrics.
+       * @default false
        */
-      deduplicationEnabled?: boolean
```
Comment (benasher44) [inline index.d.ts:452]: seems like this comma should have been required before 🤔

**Comment by benasher44** [comment]
Okay I've re-reviewed the diff a few more times to try to ensure as little outside of the restructuring that's needed has been (aside from a stray missing comma I called out).

**Comment by benasher44** [comment]
Before this change, output of [attw](https://github.com/arethetypeswrong/arethetypeswrong.github.io/tree/main/packages/cli):

image

After this change, output of attw:

image

Results were gotten by running `attw` on the result of `npm pack` on master and then again on the result of running `npm pack` on this PR. So at least on that front, this PR accomplish its type compatibility goals.

File: index.d.ts [hunk 6]
```diff
@@ -1,22 +1,27 @@
...
-
 /**
  * Tracer is the entry-point of the Datadog tracing implementation.
  */
-export declare interface Tracer extends opentracing.Tracer {
+interface Tracer extends opentracing.Tracer {
+  /**
+   * Add tracer as a named export
+   */
+  tracer: Tracer;
```
Comment (benasher44) [inline index.d.ts:13]: caught by `yarn type:test`

**Comment by benasher44** [comment]
> I worry a bit that so much structural changes might make this need to be a semver major change.

AFAICT, test coverage should confirm that these type updates are compatible, but I'm happy to open a PR to first improve coverage, if that'd be more comfortable. There are some uncovered exports, but it looks it looks like the major ones are covered by docs/test.ts:

```ts
import ddTrace, { tracer, Tracer, TracerOptions, Span, SpanContext, SpanOptions, Scope, User } from '..';
import type { plugins } from '..';
import { opentelemetry } from '..';
```

**Comment by benasher44** [comment]
Also happy to make myself available for quick follow up fixes if I'm wrong 🤞

**Review by Qard — APPROVED** [review]

**Comment by benasher44** [comment]
🙌  thanks!

For the failures: many are HTTP errors — maybe it being my commit (so my user?) or something is causing an issue? I'm not sure whether the others are expected.

Let me know, and I'm happy to update with master, if needed.

**Comment by benasher44** [comment]
Hi @Qard! I wanted to check in to see what you might need from me to be able to merge this, or if there might be something else you all are waiting on internally?

**Comment by benasher44** [comment]
Is it a confidence issue? If so, I'd be happy to open a PR to improve type test coverage first.

**Comment by Cellule** [comment]
I tested real quick on my setup doing a dual cjs/esm build and this change fixes my problem for the esm build

**Pull request closed without merging**